packages/block-editor/src/components/colors/use-colors.js

```diff
@@ -0,0 +1,140 @@
+ * block attribute setters for a given list of color types and
+ * a custom color palette.
+ *
+ * @example
+ * Calling it with a color type `"backgroundColor"` will return an object
+ * with a color object `backgroundColor`, and a function `setBackgroundColor`.
+ * `setBackgroundColor` can be called with any color value. If the value
+ * equals the `color` property of an object in the `colorPalette`,
+ * `backgroundColor` will be set to that object merged with a custom
+ * `class` property. If it's not, it will be set to a custom object
+ * with just a `color` property set to the value.
+ *
+ * ```jsx
+ * const CUSTOM_COLORS = [
+ * 	{ name: 'Red', slug: 'red', color: '#ff0000' },
+ * 	{ name: 'Blue', slug: 'blue', color: '#0000ff' },
+ * ];
+ *
+ * const MyColorfulComponent = () => {
+ * 	const { backgroundColor, setBackgroundColor } = useCustomColors(
+ * 		[ 'backgroundColor' ],
+ * 		CUSTOM_COLORS
+ * 	);
+ * 	// setBackgroundColor( '#ff0000' );
+ * 	// _.isEqual(
+ * 	// 	{ ...CUSTOM_COLORS[ 0 ], class: 'has-red-background-color' },
+ * 	// 	backgroundColor
+ * 	// );
+ * 	// setBackgroundColor( '#000000' );
+ * 	// _.isEqual( { color: '#000000' }, backgroundColor );
+ * 	//...
+ * };
+ * ```
+ *
+ * @param {(Object|string)[]} colorTypes The arguments can be strings or objects. If the argument is an object,
+ *                                        it should contain the color attribute name as key and the color context as value.
+ *                                        If the argument is a string, the value should be the color attribute name,
+ *                                        the color context is computed by applying a kebab case transform to the value.
+ *                                        Color context represents the context/place where the color is going to be used.
+ *                                        The class name of the color is generated using 'has' followed by the color name
+ *                                        and ending with the color context all in kebab case e.g: has-green-background-color.
+ *
+ * @param {Object[]} colorPalette The array of color objects (name, slug, color, etc... ).
+ *
+ * @return {Object} The object with color objects and color setters.
+ */
+export function useCustomColors( colorTypes, colorPalette ) {
+	const attributePicker = useAttributePicker( {
+		names: colorTypes.flatMap( ( colorType ) =>
+			typeof colorType === 'string' ? colorType : Object.keys( colorType )
+		),
+		valuesEnum: colorPalette,
+
+		findEnumValue: ( colorPaletteObject, newColorValue ) =>
+			colorPaletteObject.color === newColorValue,
+		mapEnumValue: ( colorPaletteObject ) => colorPaletteObject.slug,
+
+		mapAttribute: ( colorSlugOrCustomValue, name ) => {
+			const foundColorPaletteObject = colorPalette.find(
+				( colorPaletteObject ) => colorPaletteObject.slug === colorSlugOrCustomValue
+			) || { color: colorSlugOrCustomValue };
+
+			let colorClass;
+			if ( foundColorPaletteObject.slug ) {
```

Comment:
This function seems complex mainly because of retriving the colorContextName. Would something like this work?
```
const foundColorType = find( colorTypes, ( colorType ) => !! colorType && !! colorType[ name ] );
const colorContextName = foundColorType[ name ] || name;
```

Given that the mapping may execute frequently we may also consider computing an object based on  colorTypes whose keys are the name and whose value is the color context or even the class directly. The names passed to useAttributePicker would then be just the keys of that object.

Reply:
Yeah something in that style does read way better.

>Given that the mapping may execute frequently we may also consider computing an object based on colorTypes whose keys are the name and whose value is the color context or even the class directly. The names passed to useAttributePicker would then be just the keys of that object.

Also might be a good idea.